src/FSharpLint.Client/LSPFSharpLintService.fs

```diff
@@ -0,0 +1,259 @@
+                                    let createDaemonResult = createFor startInfo
+
+                                    match createDaemonResult with
+                                    | Ok daemon ->
+                                        replyChannel.Reply(Ok daemon.RpcClient)
+
+                                        { Daemons = Map.add version daemon state.Daemons
+                                          FolderToVersion = Map.add folder version state.FolderToVersion }
+                                    | Error pse ->
+                                        replyChannel.Reply(Error(GetDaemonError.FSharpLintProcessStart pse))
+                                        state
+                                | Error FSharpLintToolError.NoCompatibleVersionFound ->
+                                    replyChannel.Reply(Error GetDaemonError.InCompatibleVersionFound)
+                                    state
+                                | Error(FSharpLintToolError.DotNetListError dotNetToolListError) ->
+                                    replyChannel.Reply(Error(GetDaemonError.DotNetToolListError dotNetToolListError))
+                                    state
+                        | Reset replyChannel ->
+                            Map.toList state.Daemons
+                            |> List.iter (fun (_, daemon) -> (daemon :> IDisposable).Dispose())
+
+                            replyChannel.Reply()
+                            ServiceState.Empty
+
+                    return! messageLoop nextState
+                }
+
+            messageLoop ServiceState.Empty),
+        cancellationToken = ct
+    )
+
+type FSharpLintServiceError =
+    | DaemonNotFound of GetDaemonError
+    | FileDoesNotExist
+    | FilePathIsNotAbsolute
+    | CancellationWasRequested
+
+let isPathAbsolute (path: string) : bool =
+    if
+        String.IsNullOrWhiteSpace path
+        || path.IndexOfAny(Path.GetInvalidPathChars()) <> -1
+        || not (Path.IsPathRooted path)
+    then
+        false
+    else
+        let pathRoot = Path.GetPathRoot path
+        // Accepts X:\ and \\UNC\PATH, rejects empty string, \ and X:, but accepts / to support Linux
+        if pathRoot.Length <= 2 && pathRoot <> "/" then
+            false
+        else if pathRoot.[0] <> '\\' || pathRoot.[1] <> '\\' then
+            true
+        else
+            pathRoot.Trim('\\').IndexOf('\\') <> -1 // A UNC server name without a share name (e.g "\\NAME" or "\\NAME\") is invalid
+
+let private isCancellationRequested (requested: bool) : Result<unit, FSharpLintServiceError> =
+    if requested then
+        Error FSharpLintServiceError.CancellationWasRequested
+    else
+        Ok()
+
+let private getFolderFor filePath (): Result<Folder, FSharpLintServiceError> =
+    let handleFile filePath =
+        if not (isPathAbsolute filePath) then
+            Error FSharpLintServiceError.FilePathIsNotAbsolute
+        else match Folder.from filePath with
```

Comment:
@MrLuje see the code here above? it is feeding a "filePath" to the `Folder` type. So is it a file or a folder? if it's a folder, it should be renamed from filePath to dirPath.

Reply:
It's a folder (getting the folder of a file)

Reply:
@MrLuje ok this is confusing, I thought the point of the `from` function was to convert a path(string) to a proper type. So, in my mind, Folder.from would map a `(folderPath:string)` to a `(folder:Folder)`, and so that's why we included a `File` type equivalent to the `Folder` type that has a `From` method that works this way.

If we want to get the folder where a file lives, I think we should create an API that is more similar to the already existing BCL's System.IO API, to avoid confusions. So, in System.IO, the way to do this is convert the `filePath:string` to a `FileInfo`, and then accessing the `.Directory` property. So how about changing `Folder`'s `From` method to receive a folderPath (not filePath; when given a filePath it should error saying that it is a file, not a folder). So if in this piece of code you want to get the folder where certain path lives, just create it via `File.From` and then let's create a `Folder` property in `File` type?

Reply:
I fear that it might be weird from a discovery POV (signature is a `Folder` but no clue that you have to go through a `File` before)
What about if `Folder.From` can handle both a filePath and a folderPath ? or a `FromFile` / `FromFolder` ?

Reply:
@MrLuje I like the FromFile/FromFolder idea, let's do it

Reply:
Done
I missed it when changing to `FileInfo` but I don't need `Path.GetFullPath` anymore, `FileInfo` handles it properly